lib/Index/Index.cpp

```diff
@@ -216,26 +216,38 @@ class ContainerTracker {
 
   bool empty() const { return Stack.empty(); }
 
-  void forEachActiveContainer(llvm::function_ref<void(const Decl *)> f) const {
-    if (Stack.empty())
-      return;
+  void forEachActiveContainer(llvm::function_ref<bool(const Decl *)> allowDecl,
+                              llvm::function_ref<void(const Decl *)> f) const {
+    for (const auto &Entry : llvm::reverse(Stack)) {
+      // No active container, we're done.
+      if (!Entry.ActiveKey)
+        return;
 
-    const StackEntry &Entry = Stack.back();
+      auto MapEntry = Entry.Containers.find(Entry.ActiveKey);
+      if (MapEntry == Entry.Containers.end())
+        return;
 
-    if (!Entry.ActiveKey)
-      return;
+      SmallVector<const Decl *, 4> Containers;
```

Comment:
Could have a bool for whether we actually allowed any decls or not and then just use `f(D)` in the two if's, but if you prefer this I'd move the `Containers` out of the loop.

Reply:
Sure, don't feel too strongly, switched to using a bool